{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How close are Codon and Amino Acid embeddings of the same sequence?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Start with visualization\n",
    "Let's look at t-SNE of (1) the codon embeddings, (2) the amino acid embeddings, and (3) the combined embeddings.\n",
    "\n",
    "### First generate embeddings for codon and amino acid sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Implement JIT compile model: https://huggingface.co/docs/transformers/torchscript\n",
    "def generate_embeddings(model, dataloader):\n",
    "    embeddings = []\n",
    "    with torch.no_grad():\n",
    "        for batch in tqdm(dataloader):\n",
    "            batch = batch.to(model.device)\n",
    "            outputs = model(**batch, output_hidden_states=True)\n",
    "            last_hidden_states = outputs.hidden_states[-1]\n",
    "            seq_lengths = attention_mask.sum(axis=1)\n",
    "            for seq_len, elem in zip(seq_lengths, last_hidden_states):\n",
    "                embedding = elem[1 : seq_len - 1, :].cpu().numpy()\n",
    "                embeddings.append(embedding)\n",
    "\n",
    "    return np.array(embeddings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dataclasses import dataclass\n",
    "from torch.utils.data import DataLoader\n",
    "from transformers import (\n",
    "    PreTrainedTokenizerFast,\n",
    "    BatchEncoding,\n",
    "    DataCollatorForLanguageModeling,\n",
    "    Trainer,\n",
    "    TrainingArguments,\n",
    "    EsmForMaskedLM,\n",
    "    EsmTokenizer,\n",
    ")\n",
    "\n",
    "from train import GenSLMColatorForLanguageModeling, FastaDataset\n",
    "\n",
    "@dataclass\n",
    "class GenSLMInferenceConfig:\n",
    "    # Could also be an ESM checkpoint, e.g. \"esm1_t34_670M_UR50S\"\n",
    "    model_path: str = \"facebook/esm2_t6_8M_UR50D\"\n",
    "    tokenizer_path: str = \"tokenizer_esm_genslm\"\n",
    "    data_path: str = \"/lambda_stor/homes/khippe/genslm_foundation/genome_data/mdh_sc23/fasta/mdh_natural_sequences.ffn\"\n",
    "    per_device_eval_batch_size: int = 128\n",
    "\n",
    "config = GenSLMInferenceConfig()\n",
    "\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "tokenizer = EsmTokenizer.from_pretrained(config.tokenizer_path)\n",
    "model = EsmForMaskedLM.from_pretrained(config.model_path)\n",
    "model.to(device).eval()\n",
    "\n",
    "dataset = FastaDataset(   \n",
    "    file_path=config.data_path,\n",
    "    return_codon=True,\n",
    "    return_aminoacid=True,\n",
    ")\n",
    "\n",
    "data_collator = GenSLMColatorForLanguageModeling(\n",
    "    return_codon=True,\n",
    "    return_aminoacid=False,\n",
    "    tokenizer=tokenizer,\n",
    ")\n",
    "\n",
    "dataloader = DataLoader(\n",
    "    dataset,\n",
    "    batch_size=config.per_device_eval_batch_size,\n",
    "    collate_fn=data_collator,\n",
    "    num_workers=1,\n",
    "    pin_memory=True\n",
    ")\n",
    "\n",
    "embeddings = generate_embeddings(model, dataloader)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Investigate whether the Codon and AA embeddings are within the top-k nearest neighbors\n",
    "1. Embed codon sequences\n",
    "2. Embed AA sequences\n",
    "3. Find the top-k nearest neighbors for each sequence\n",
    "4. For each codon sequence, count the number of times the AA sequence is in the top-k nearest neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
